File: C++/Problems/Problems/problem_2294.cpp
#include "problem_2294.h"

#include <algorithm>
#include <iostream>
#include <ostream>

bool problem_2294::test() {
    std::vector nums { 3, 6, 1, 2, 5 };

    auto k = 2;

    auto expected = 2;

    auto result = partitionArray(nums, k);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2294::partitionArray(std::vector<int> &nums, int k) {
    std::sort(begin(nums), end(nums));
    auto cnt = 0;
    auto prev = nums[0];
    auto result = 0;
    auto addLast = true;
    for (int i = 0; i < nums.size(); i++)
    {
        if (nums[i] != prev)
        {
            addLast = false;
            cnt += nums[i] - prev;
            prev = nums[i];
            if (cnt > k)
            {
                cnt = 0;
                result++;
                addLast = true;
            }
        }
    }

    if (cnt != 0 || addLast)
        result++;

    return result;
}

File: Python/Problems/Problems/problem2294.py
from .problem import Problem

class Problem2294(Problem):
    def test(self):
        nums = [3, 6, 1, 2, 5]

        k = 2

        expected = 2

        result = self.partitionArray(nums, k)

        print(result)

        return result == expected

    def partitionArray(self, nums, k):
        nums.sort()
        cnt = 0
        prev = nums[0]
        result = 0
        addLast = True
        for i in range(len(nums)):
            if nums[i] != prev:
                addLast = False
                cnt += nums[i] - prev
                prev = nums[i]
                if cnt > k:
                    cnt = 0
                    result += 1
                    addLast = True

        if cnt != 0 or addLast:
            result += 1

        return result